_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,740 @@
+class of data via `enum`, and processing instances of such data via
+`match`.
+
+Most of the post will use a single running example: modeling a game of
+chess. At the end the post changes gears, shifting to a discussion of
+how affine-typing ensures that the demonstrated features do not
+introduce unsoundness.
+
+## Tutorial material
+
+The post takes the overall form of a tutorial.
+
+If you are already familiar with defining `enums` and simple uses of
+`match`, you may want to skip ahead to the [mutating enums] or
+[avoiding unsoundness] sections. This first section is meant to
+establish everything one would need to know about enums to understand
+the discussion in latter two sections.
+
+[mutating enums]: #mutating-enums
+[avoiding unsoundness]: #avoiding-unsoundness
+
+### Defining simple enums
+
+The pieces of chess can be classified as follows:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PieceShape {
+    King, Queen, Rook, Bishop, Knight, Pawn,
+}
+```
+
+The `derive` line automatically generates support for three common
+operations: the `Copy, Clone` gives us the ability to freely copy or
+clone instances of `PieceShape`, the `Debug` provides support for
+printing each piece-rank into an output buffer, e.g. via
+`println!("piece: {:?}", piece)`, and `PartialEq, Eq` provides support
+for the `==` operator on instances of the enum.
+
+### Enum discriminant values
+
+An enum like `PieceShape` above, where each variant is just a name, is
+represented using an integer value for each variant. The definition
+style above allows the Rust compiler to assign the values itself; it
+will start from 0 and count upwards.
+
+Alternatively, one can manually assign specific values to enum
+variants, interrupting the automatic value assignment performed by the
+compiler. Thus, we could instead have written the below, and get the
+same assignment of discriminant values:
+
+```rust
+#[cfg(alternative)]
+#[derive(Copy, Clone, Debug)]
+pub enum PieceShape {
+    Pawn = 5, Rook = 2, Bishop, Knight, King = 0, Queen, 
+}
+```
+
+(In the above, the `cfg` line is how one denotes "conditional
+compilation" in Rust; it indicates that the `enum` definition beneath
+it should only be compiled if one passes `--cfg alternative` to the
+invocation of `rustc`; it is being used above as a quick-and-dirty way
+to avoid causing a compile-time error with a duplicate definition.)
+
+These enums, where each variant is just a name with an associated
+(implicitly- or explicitly-assigned) integer value, are known as
+"C-style enums" in Rust parlance, since the syntax largely matches
+that used for the `enum` construct in the C language.
+
+(However, note that not every C enum immediately corresponds to an
+analogous enum in Rust; in particular, Rust enforces an invariant that
+each variant be distinct from all other variants in the enum, while C
+allows one to assign the same integer value to multiple enum
+variants.)
```

Comment:
This sentence about C enum correspondence seems a bit much to be included within the main portion of the tutorial.  Perhaps it should be turned into a footnote instead?